File: backend-node/src/services/feeCalculator.ts
import { Money } from '../domain/money.js';

/**
 * Port of Services/FeeCalculator.cs.
 *
 * MDR = 0.7% + Rp 500 flat, always FLOORED to a whole rupiah — never
 * Math.round, never MidpointRounding.AwayFromZero. The C# original uses
 * `decimal.Floor`; here the same guarantee is achieved with exact integer
 * arithmetic (units * 7 / 1000, floored) so no floating-point multiply
 * ever touches a stored amount.
 *
 * NOTE — intentional gap, same as the .NET original: there is no
 * `calculateFeeReversal` for partial refunds. Specifying and implementing
 * that formula (floor(fee_charged × refund_amount ÷ gross)) is Lab 1/2's
 * job, not something this file does for you.
 */
export const MDR_NUMERATOR = 7; // 0.7% == 7 / 1000
export const MDR_DENOMINATOR = 1000;
export const FIXED_FEE_UNITS = 500;

export function calculateMerchantFee(gross: Money): Money {
  const percentagePart = Math.floor((gross.units * MDR_NUMERATOR) / MDR_DENOMINATOR);
  return Money.idr(percentagePart + FIXED_FEE_UNITS);
}

File: backend/src/PaymentLab.Api/Services/FeeCalculator.cs
using PaymentLab.Api.Domain;

namespace PaymentLab.Api.Services;

/// <summary>
/// Merchant fee (MDR) arithmetic.
///
/// CONVENTION FOR THIS WHOLE CLASS: every fee amount is floored to a whole rupiah.
/// Flooring is deliberate — the PJP must never charge, and never reverse,
/// a fraction more than the exact figure. Do not "improve" this to Math.Round.
/// </summary>
public sealed class FeeCalculator
{
    /// <summary>Merchant discount rate: 0.7% of the principal.</summary>
    public const decimal MdrRate = 0.007m;

    /// <summary>Flat per-transaction fee, in IDR.</summary>
    public const long FixedFeeUnits = 500;

    public Money CalculateMerchantFee(Money gross)
    {
        if (gross.IsNegative)
        {
            throw new ArgumentOutOfRangeException(nameof(gross), "Gross amount must not be negative.");
        }

        var variable = (long)decimal.Floor(gross.Units * MdrRate);
        return new Money(variable + FixedFeeUnits, gross.Currency);
    }
}
